microbenchmarks/singledispatch.py

```diff
@@ -0,0 +1,53 @@
+from functools import singledispatch
+
+from benchmarking import benchmark
+
+NUM_ITER = 100_000
+
+
+class Tree:
+    pass
+
+
+class Leaf(Tree):
+    pass
+
+
+class Node(Tree):
+    def __init__(self, value: int, left: Tree, right: Tree) -> None:
+        self.value = value
+        self.left = left
+        self.right = right
+
+
+@singledispatch
+def calc_sum(x: Tree) -> int:
+    raise TypeError("invalid type for x")
+
+
+@calc_sum.register
```

Comment:
I get an exception here when compiled:
```
TypeError: Invalid first argument to `register()`: <built-in function __mypyc_sum_leaf_decorator_helper__>. Use either `@register(some_class)` or plain `@register` on an annotated function.
```

If the benchmark doesn't work on mypyc yet, we probably shouldn't merge this PR, as we won't be able to collect any results. 

One option would be to make it possible to mark this as 'skipped' benchmark and have the benchmark server scripts skip this benchmark until it's known to work.

Reply:
This should be fixed in c878e2e3ff17792c183709cfb3beb7d68336620d. I think the problem is that compiled functions don't appear to preserve `__annotations__`, which `register` is probably trying to use to get the type of the first argument. Passing the class as an argument to `register` seems to work for now, but special casing `singledispatch` should make that unnecessary.